From 4d387173336e013d3d0389e652847ac02b508275 Mon Sep 17 00:00:00 2001
From: Suneel Garapati <suneel.garapati@cavium.com>
Date: Wed, 11 Jan 2017 17:54:23 -0800
Subject: [PATCH 0595/1239] cmd: bootimgup: validate boot image before update

use cavium flash header and bdk magic string to validate
image before flashing to device.

Signed-off-by: Suneel Garapati <suneel.garapati@cavium.com>
---
 cmd/bootimgup.c | 20 ++++++++++++++++++++
 1 file changed, 20 insertions(+)

diff --git a/cmd/bootimgup.c b/cmd/bootimgup.c
index d26a2dd522..f120a907b5 100644
--- a/cmd/bootimgup.c
+++ b/cmd/bootimgup.c
@@ -21,6 +21,21 @@
 
 static struct spi_flash *flash;
 
+static int validate_bootimg_header(unsigned long addr)
+{
+	char flash_hdr[] = {"CVM_CLIB"};
+	char bdk_magic[] = {"THUNDERX"};
+	char *buf1 = (char *)(addr + 0x10000); /* flash hdr offset */
+	char *buf2 = (char *)(addr + 0x20008); /* bdk magic offset */
+	char *buf3 = (char *)(addr + 0x50008); /* sec bdk magic offset */
+
+	if ( strncmp(buf1, flash_hdr, 8) == 0 )
+		if ( strncmp(buf2, bdk_magic, 8) == 0 )
+			if (strncmp (buf3, bdk_magic, 8) == 0 )
+				return 0;
+	return 1;
+}
+
 /**
  * This function takes a byte length and a delta unit of time to compute the
  * approximate bytes per second
@@ -209,6 +224,11 @@ static int do_bootu_spi(int argc, char * const argv[])
 		debug(" %s len %ld\n", __func__, len);
 	}
 
+	if (validate_bootimg_header(addr)) {
+		printf("\n No valid boot image header found \n");
+		return 1;
+	}
+
 	/* The remaining commands require a selected device */
 	if (!flash) {
 		debug("No SPI flash selected.  Executing 'sf probe'\n");
-- 
2.29.0

